boltstub/tests/test_wiring.py

```diff
@@ -0,0 +1,341 @@
+
+from functools import reduce
+from random import getrandbits
+
+import pytest
+
+from ..wiring import (
+    create_wire,
+    negotiate_socket,
+    RegularSocket,
+    WebSocket,
+)
+
+
+def randbytes(size):
+    return bytearray(getrandbits(8) for _ in range(size))
+
+
+class TestRegularSocket:
+    class TestWhenCacheIsSupplied:
+
+        def test_recv_should_return_cache(self, mocker):
+            cache = b"abc"
+
+            socket_mock = mocker.Mock()
+            regular_socket = RegularSocket(socket_mock, cache)
+
+            received = regular_socket.recv(1024)
+
+            assert received == cache
+            socket_mock.recv.assert_not_called()
+
+        def test_accessing_the_socket_during_second_call(self, mocker):
+            cache = b"abc"
+            expected = b"barcelos"
+
+            socket_mock = mocker.Mock()
+            socket_mock.recv.return_value = expected
+            regular_socket = RegularSocket(socket_mock, cache)
+
+            regular_socket.recv(1024)
+            actual = regular_socket.recv(1024)
+
+            assert actual == expected
+            socket_mock.recv.assert_called_with(1024)
+
+    class TestWhenCacheIsEmpty:
+
+        @pytest.mark.parametrize("empty_cache", [
+            (None),
+            (b"")
+        ])
+        def test_accessing_the_socket(self, empty_cache, mocker):
+            expected = b"barcelos"
+
+            socket_mock = mocker.Mock()
+            socket_mock.recv.return_value = expected
+            regular_socket = RegularSocket(socket_mock, empty_cache)
+
+            actual = regular_socket.recv(1024)
+
+            assert actual == expected
+            socket_mock.recv.assert_called_with(1024)
+
+    @pytest.mark.parametrize("method_name", [
+        ("close"),
+        ("send"),
+        ("sendall"),
+        ("settimeout"),
+        ("getsockname"),
+        ("getpeername")
+    ])
+    def test_proxying_sockets_members(self, method_name, mocker):
+        socket_mock = mocker.Mock()
+        regular_socket = RegularSocket(socket_mock, None)
+
+        wrapped_method = getattr(regular_socket, method_name)
+        original_method = getattr(socket_mock, method_name)
+
+        assert wrapped_method == original_method
+
+
+def duplicate(obj, times):
+    if times == 1:
+        return obj
+    return duplicate(obj + obj, times - 1)
```

Comment:
FYI: function calls in Python are relatively costly. You'd probably be more efficient with a loop or a reduce.

```python
for _ in range(times):
    obj += obj
```

or 

```python
from functools import reduce
from operator import add

reduce(add, (obj for _ in range (times)))
```

Now, since this is only for testing and times will (probably) not be huge, it won't really make a difference. Just wanted to mention it. Also, Python limits the depth of call stacks by default to something like 1000 or so. So recursion isn't generally my go-to paradigm in Python unless it makes things much easier.

However, I can't find, where this function is used anyway?

Reply:
This function should be a leftover of previous implementation of some test and I was probably trying to duplicate something 32 or 64 times, or something like this. It was probably for generating long byte arrays.

Reply:
Function calls in JS is almost cost free, most of time it is translated to inline code. The same happens with Java and C++ (in this case, you have to give to the compiler some hints).